### DISCUSSION TITLE
How to build solid directly from non-planar faces?

### DISCUSSION THREAD

**Discussion by GarryBGoode** [CadQuery/cadquery#1561] [Q&A]
Hello!

I'm working on importing manim objects to CadQuery.
Manim objects are built from 4-point Bezier curve segments, so after figuring out how CQ likes its b-splines, it should be possible and not too hard...

My problem is that I can't find a way to directly build a solid from non-planar surface elements.
I tried constructing Edges and then a Wire from those, but then the Face class won't let me create it for not being planar :(

Here's an example code (needs manim):

```
import manim as mnm
import cadquery as cq
import numpy as np

cyl = mnm.Cylinder()

edges = []
for k in range(0,cyl.submobjects[0].get_num_curves()):
    points = cyl.submobjects[0].points[4*k:4*k+4,:]
    vectors = [cq.Vector((points[k,0],points[k,1],points[k,2])) for k in range(4)]
    tangent_vals=[(points[1,:]-points[0,:])/np.linalg.norm(points[1,:]-points[0,:]),
                  (points[3,:]-points[2,:])/np.linalg.norm(points[3,:]-points[2,:])]
    tangents = [cq.Vector((p[0],p[1],p[2])) for p in tangent_vals]

    edges.append(cq.Edge.makeSpline([vectors[0],vectors[3]],tangents, scale=False))

result = cq.Wire.assembleEdges(edges)

show_object(result)

result = cq.Face.makeFromWires(result)
```

**Comment by michaelgale** [comment]
While I can't provide a working implementation of what you're trying to achieve, I would consult the [`makeSplineApprox`](https://cadquery.readthedocs.io/en/latest/classreference.html#cadquery.Face) classmethod in the `Face` class or the [`interpPlate`](https://cadquery.readthedocs.io/en/latest/classreference.html#cadquery.Solid) classmethod in the `Solid` class.  The `Face` method can be used to generate multiple faces which can then be stitched together into a `Shell` and then into a `Solid`.  The `interpPlate` method can directly yield a solid object representing a finite thickness plate.

At quick glance it appears the `interpPlate` method can be provided with either a list of edges or a list of 3D points.  This method generates a solid which best approximates the surface described by the edges or points.

**Comment by GarryBGoode** [comment]
Thanks!
interpPlate looks promising, allegedly it returns a face when the thickness is 0.
Just for some reason I can't feed it this list of Edges, nor the assembled Wire, nor a list of Wires.

I'm also not a fan of interpolations when I already have spline patches. I know under the hood everything must be some form of spline patches... just lemme at 'em.
In the end I might need to evaluate a number of points from the manim splines, feed the points to the other approximator and hope the result is close enough.

And it's got a deprecation warning, interpPlate will be removed. I wonder what's up with that?

**Comment by adam-urbanczyk** [comment]
What is your real question? If you want to build a solid from faces, use `Shell.makeShell` and then `Solid.makeSolid`.

**Comment by GarryBGoode** [comment]
In the end I want to build a Solid with curved surfaces from splines that I calculate myself. The actual end goal is to make a gear generator.
Yes I wanted to make Faces and then Shell and then Solid.

As of yet I can't make a Face.

`Face.makeFromWires()` is complaining it's not planar. The `interpPlate` method is a bit confusing, the documentation says it can take a list of edges or wires but I haven't been able to make it work with wires.

If you can provide an example where `interpPlate` works on a wire or edges that would be great.

I am hesitant to use the approximation methods, eg. using `interpPlate` with points or `makeSplineApprox` cause I want full control over the shape, and I would like to avoid reverse engineering the quirks of the approximation.

**Reply by adam-urbanczyk** [reply]
Ok, the title of the post does not reflect your question then. Do you actually have the B-spline surface, or do you only have edges?

In the first case you'll need to use OCP/OCCT directly. In the second case you can use indeed `interplate` or better `Face.makeNSidedSurface`.  https://github.com/CadQuery/cadquery/blob/3451007f8eeb9d78e784ec8047ef69a5359ecb5e/tests/test_cadquery.py#L5587

**Comment by GarryBGoode** [comment]
OK, feels like I'm almost there. I was able to make the surfaces with `Face.makeNSidedSurface`. Then I converted them to a shell with `Shell.makeShell`.

Still, Solid.makeSolid does not take the resulting shell from the previous operations.
Visually it looks OK, so I don't know what goes wrong here.
```
TypeError: SolidFromShell(): incompatible function arguments. The following argument types are supported:
    1. (self: OCP.ShapeFix.ShapeFix_Solid, shell: OCP.TopoDS.TopoDS_Shell) -> OCP.TopoDS.TopoDS_Solid
  ```
![image](https://github.com/CadQuery/cadquery/assets/95878914/4b354aab-f5df-4fff-93ab-9778d91468af)

```import manim as mnm
import cadquery as cq
import numpy as np

cyl = mnm.Cylinder()
faces = []
for mob in cyl.family_members_with_points():
    edges = []
    for k in range(0,mob.get_num_curves()):
        points = mob.points[4*k:4*k+4,:]
        vectors = [cq.Vector((points[k,0],points[k,1],points[k,2])) for k in range(4)]
        tangent_vals=[(points[1,:]-points[0,:])/np.linalg.norm(points[1,:]-points[0,:]),
                      (points[3,:]-points[2,:])/np.linalg.norm(points[3,:]-points[2,:])]
        tangents = [cq.Vector((p[0],p[1],p[2])) for p in tangent_vals]
        edges.append(cq.Edge.makeSpline([vectors[0],vectors[3]],tangents, scale=False))
    
    faces.append(cq.Face.makeNSidedSurface(edges,[]))
    
result = cq.Shell.makeShell(faces)
show_object(result)
result = cq.Solid.makeSolid(result)
```

**Comment by GarryBGoode** [comment]
OK, so I guess I found the answer, back here to report...

The issue was more with the topology, the top and bottle circle had 8 segments, the sides had something like 24 segments around the circular surface.
I think the kernel could not stitch it together into a nice shell.
When I changed the sides to be low res and have only 8 segments tangentially, it worked.
